packages/arb-token-bridge-ui/src/hooks/useSelectedToken.ts

```diff
@@ -0,0 +1,198 @@
+import { useCallback } from 'react'
+import { utils } from 'ethers'
+import useSWRImmutable from 'swr/immutable'
+import { Provider } from '@ethersproject/providers'
+import {
+  getChainIdFromProvider,
+  getProviderForChainId
+} from '@/token-bridge-sdk/utils'
+
+import { ERC20BridgeToken, TokenType } from './arbTokenBridge.types'
+import {
+  getL2ERC20Address,
+  isTokenArbitrumOneNativeUSDC,
+  isTokenArbitrumSepoliaNativeUSDC,
+  isTokenMainnetUSDC,
+  isTokenNativeUSDC,
+  isTokenSepoliaUSDC
+} from '../util/TokenUtils'
+import { useNetworks } from './useNetworks'
+import { isNetwork } from '../util/networks'
+import { CommonAddress } from '../util/CommonAddressUtils'
+import { useNetworksRelationship } from './useNetworksRelationship'
+import {
+  useTokensFromLists,
+  useTokensFromUser
+} from '../components/TransferPanel/TokenSearchUtils'
+import { useArbQueryParams } from './useArbQueryParams'
+import { useTokenLists } from './useTokenLists'
+
+const commonUSDC = {
+  name: 'USD Coin',
+  type: TokenType.ERC20,
+  symbol: 'USDC',
+  decimals: 6,
+  listIds: new Set<string>()
+}
+
+export const useSelectedToken = () => {
+  const [{ token: tokenFromSearchParams }, setQueryParams] = useArbQueryParams()
+  const [networks] = useNetworks()
+  const { childChain, parentChain } = useNetworksRelationship(networks)
+  const tokensFromLists = useTokensFromLists()
+  const tokensFromUser = useTokensFromUser()
+  const { isLoading: isLoadingTokenLists } = useTokenLists(childChain.id)
+
+  const queryKey =
+    !isLoadingTokenLists &&
+    tokenFromSearchParams &&
+    tokensFromLists &&
+    tokensFromUser
+      ? ([
+          parentChain.id,
+          childChain.id,
+          tokenFromSearchParams,
+          JSON.stringify(tokensFromLists),
+          JSON.stringify(tokensFromUser),
+          'useSelectedToken'
+        ] as const)
+      : null
+
+  const { data } = useSWRImmutable(
+    queryKey,
+    async ([
+      parentChainId,
+      childChainId,
+      _tokenFromSearchParams,
+      _tokensFromLists,
+      _tokensFromUser
+    ]) => {
+      const parentProvider = getProviderForChainId(parentChainId)
+      const childProvider = getProviderForChainId(childChainId)
+
+      if (isTokenNativeUSDC(_tokenFromSearchParams)) {
+        return getUsdcToken({
+          tokenAddress: _tokenFromSearchParams,
+          parentProvider,
+          childProvider
+        })
+      }
+
+      return (
+        JSON.parse(_tokensFromLists)[_tokenFromSearchParams] ||
+        JSON.parse(_tokensFromUser)[_tokenFromSearchParams] ||
+        null
+      )
+    }
+  )
```

Comment:
NIT: I think it would make more sense to return the address in `useSWRImmutable` without `tokensFromLists` and `tokensFromUser`, and have a simple `useMemo` to get the data
```
const token = useMemo(() => {
  return tokensFromLists[memoedAddressFromSWR] || tokensFromUser[memoedAddressFromSWR] || null 
}, [tokensFromLists, tokensFromUser)
```